<?xml version="1.0"?>

<robot name="limo_four_diff" xmlns:xacro="http://ros.org/wiki/xacro">

    <xacro:include filename="$(find lab1_mobile_ackermann)/urdf/limo_xacro.xacro" />
    <xacro:include filename="$(find lab1_mobile_ackermann)/urdf/limo_steering_hinge.xacro" />

    <!-- Variables -->
    <xacro:property name="M_PI" value="3.14159"/>

    <!-- Vehicle Geometries -->
    <xacro:property name="base_x_size" value="0.13" />
    <xacro:property name="base_y_size" value="0.12" />
    <xacro:property name="base_z_size" value="0.1" />
  
    <xacro:property name="wheelbase" value="0.2"/>
    <xacro:property name="track" value="0.13"/>
    <xacro:property name="wheel_vertical_offset" value="-0.10" />
    <xacro:property name="base_mass" value="2.1557"/>

    
    <!-- <xacro:property name="wheel_length" value="0.045" /> -->
    <xacro:property name="wheel_length" value="0.001" />
    <xacro:property name="wheel_radius" value="0.045" />



  <link name="base_footprint"/>

  <joint name="base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin xyz="0.0 0.0 0.15" rpy="0 0 0"/>
  </joint>
    <!-- Base link -->
    <link name="base_link">
        <visual>
            <origin xyz="0 0 -0.15" rpy="0 0 1.57" />
            <geometry>
                <mesh filename="package://lab1_mobile_ackermann/meshes/limo_optimal.dae" scale="1 1 1"/>
                <!-- <mesh filename="file://$(find lab1_mobile_ackermann)/meshes/limo_optimal.dae" scale="1 1 1"/> -->
            </geometry>
        </visual>
        <collision>
            <!--origin xyz="0 0 -0.15" rpy="0 0 1.57" /-->
            <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry>
              <box size="${base_x_size} ${base_y_size} ${base_z_size}"/>
          </geometry>
        </collision>
    </link>

    <link name="inertial_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" />
            <mass value="${base_mass}" />
            <inertia ixx="0.0043832566666666675" ixy="0" ixz="0" 
                     iyy="0.004832360833333334" iyz="0" 
                     izz="0.005622784166666667" />
        </inertial>
    </link>

    <joint name="inertial_joint" type="fixed">
        <origin xyz="0 0 0" rpy="0 0 0" />
        <parent link="base_link" />
        <child link="inertial_link" />
    </joint>

    <xacro:limo_laser parent_prefix="base_link" frame_prefix="laser">
        <origin xyz="0.103 0 -0.034" rpy="0 0 0"/>
    </xacro:limo_laser>

    <xacro:limo_depth_camera parent_prefix="base_link" frame_prefix="depth_camera">
        <!-- <origin xyz="0.084 0 0.3" rpy="0 0 0"/> -->
        <origin xyz="0.084 0 0.026" rpy="0 0 0"/>
    </xacro:limo_depth_camera>
    
    <xacro:limo_imu parent_prefix="base_link" frame_prefix="imu">
        <origin xyz="0.0 0 -0.1" rpy="0 0 0"/>
    </xacro:limo_imu>




  <xacro:limo_left_steering_hinge parent_prefix="base_link" wheel_prefix="front_left">
    <origin xyz="${wheelbase/2} ${track/2} ${wheel_vertical_offset}" rpy="0 0  0" />
  </xacro:limo_left_steering_hinge>

  <xacro:limo_right_steering_hinge parent_prefix="base_link" wheel_prefix="front_right">
    <origin xyz="${wheelbase/2} ${-track/2} ${wheel_vertical_offset}" rpy="${M_PI} 0 0" />
  </xacro:limo_right_steering_hinge>

    <xacro:limo_wheel parent_prefix="base_link" wheel_prefix="rear_left" reflect="1" fixed_on="false">
        <origin xyz="${-wheelbase/2} ${track/2} ${wheel_vertical_offset}" rpy="0 0 0" />
    </xacro:limo_wheel>

    <xacro:limo_wheel parent_prefix="base_link" wheel_prefix="rear_right" reflect="-1" fixed_on="false">
        <origin xyz="${-wheelbase/2} ${-track/2} ${wheel_vertical_offset}" rpy="${M_PI} 0 0" />
    </xacro:limo_wheel>


    

<xacro:include filename="$(find lab1_mobile_ackermann)/urdf/limo_four_acker.gazebo" />


<gazebo>
        <plugin name="gazebo_ros_ackermann_drive" filename="libgazebo_ros_ackermann_drive.so">

        <ros>
          <namespace>demo</namespace>
          <remapping>cmd_vel:=cmd_demo</remapping>
          <remapping>odom:=odom_demo</remapping>
          <remapping>distance:=distance_demo</remapping>
          <remapping>steerangle:=steerangle_demo</remapping>
        </ros>

        <update_rate>100.0</update_rate>

        <!-- wheels -->
        <front_left_joint>front_left_wheel</front_left_joint>
        <front_right_joint>front_right_wheel</front_right_joint>
        <rear_left_joint>rear_left_wheel</rear_left_joint>
        <rear_right_joint>rear_right_wheel</rear_right_joint>
        <left_steering_joint>front_left_steering</left_steering_joint>
        <right_steering_joint>front_right_steering</right_steering_joint>
        <steering_wheel_joint>main_steering</steering_wheel_joint>

        <!-- Max absolute steer angle for tyre in radians-->
        <!-- Any cmd_vel angular z greater than this would be capped -->
        <max_steer>0.6458</max_steer>

        <!-- Max absolute steering angle of steering wheel -->
        <max_steering_angle>7.85</max_steering_angle>

        <!-- Max absolute linear speed in m/s -->
        <max_speed>20</max_speed>

        <!-- PID tuning -->
        <left_steering_pid_gain>100 0 1</left_steering_pid_gain>
        <left_steering_i_range>0 0</left_steering_i_range>
        <right_steering_pid_gain>100 0 1</right_steering_pid_gain>
        <right_steering_i_range>0 0</right_steering_i_range>
        <linear_velocity_pid_gain>0.5 0 0</linear_velocity_pid_gain>
        <linear_velocity_i_range>0 0</linear_velocity_i_range>

        <!-- output -->
        <publish_odom>true</publish_odom>
        <publish_odom_tf>true</publish_odom_tf>
        <publish_wheel_tf>true</publish_wheel_tf>
        <publish_distance>true</publish_distance>
        <publish_steerangle>true</publish_steerangle>

        <odometry_frame>odom</odometry_frame>
        <robot_base_frame>base_footprint</robot_base_frame>

      </plugin>
    </gazebo>






    
    <gazebo reference="imu_link">
    <gravity>true</gravity>
    <sensor name="twr_imu" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <imu>
        <orientation>
          <x>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-3</stddev>
            </noise>
          </x>
          <y>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-3</stddev>
            </noise>
          </y>
          <z>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-3</stddev>
            </noise>
          </z>
        </orientation>
        <angular_velocity>
          <x>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
            </noise>
          </x>
          <y>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
            </noise>
          </y>
          <z>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
            </noise>
          </z>
        </angular_velocity>
        <linear_acceleration>
          <x>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
            </noise>
          </x>
          <y>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
            </noise>
          </y>
          <z>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
            </noise>
          </z>
        </linear_acceleration>
      </imu>
      <plugin name="two_wheeled_robot_imu" filename="libgazebo_ros_imu_sensor.so">
        <initial_orientation_as_reference>false</initial_orientation_as_reference>
        <frame_name>imu_link</frame_name>
        <ros>
          <remapping>~/out:=imu</remapping>
        </ros>
      </plugin>
    </sensor>
</gazebo>

</robot>